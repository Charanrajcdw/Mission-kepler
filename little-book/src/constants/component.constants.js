export const MODAL = {
  members: "Members",
  newBlog: "Add New Blog",
  remove: "Remove",
};

export const MEMBERS_LIST = {
  title: "Members",
  noMembers: "No members available",
};

export const BLOG_FORM = {
  title: "Add New Blog",
  button: "ADD",
  titlePlaceholder: "Name your blog",
  detailsPlaceholder: "Write Content Here ..",
  invalidValues: "Invalid title or description",
  invalidBlog: "Blog title already exists",
  blogAdded: "New Blog added successfully",
  localBlog: "local",
};

export const WARNING_MODAL = {
  title: "Unsaved changes will be lost. Are you sure to continue?",
  continue: "Continue",
  cancel: "Cancel",
};
