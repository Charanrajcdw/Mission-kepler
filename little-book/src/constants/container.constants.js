export const SIDEBAR = {
  logo: "Little Book",
  filter: "FILTER",
  noFilter: "Add blogs to show filters",
  viewMembers: "View Members",
  switchToDark: "Switch to Dark Mode",
  switchToLight: "Switch to Light Mode",
  blogs: "Blogs",
};
