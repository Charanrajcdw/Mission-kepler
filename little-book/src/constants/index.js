export * from "./url.constants";
export * from "./theme.constants";
