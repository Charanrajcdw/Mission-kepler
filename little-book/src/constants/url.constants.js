export const API_URLS = {
  blogs: "https://jsonmockserver.vercel.app/api/blogs",
  users: "https://jsonmockserver.vercel.app/api/users",
  images: "https://jsonmockserver.vercel.app/",
};
