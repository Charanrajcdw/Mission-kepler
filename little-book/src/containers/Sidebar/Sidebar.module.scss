@import "../../scss/fonts";

.sidebarContainer {
  background-color: var(--primary-bg);
  width: 15.5%;
  height: 100vh;
  display: flex;
  flex-flow: column nowrap;
  justify-content: space-between;
  padding: 2.75rem 2.68rem 5.5rem;
  color: var(--primary-font-bg);
  font-family: $font-ibm;

  .title {
    font-family: $font-roboto;
    margin-left: 18px;
    font-size: 1.68rem;
    font-weight: 400;
    letter-spacing: 1px;
    color: var(--logo);

    .titleLastWord {
      color: var(--primary-font-bg);
      font-weight: 500;
    }
  }

  .filtersContainer {
    margin-top: 4.5rem;

    .filterTitle {
      margin-bottom: 20px;
      font-size: 1.18rem;
      font-weight: 500;
    }
  }

  .optionsContainer {
    font-size: 0.9rem;
    cursor: pointer;

    .theme {
      margin-top: 10px;
    }
  }
}
