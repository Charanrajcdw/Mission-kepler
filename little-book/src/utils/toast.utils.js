import { toast } from "react-toastify";

export const showSuccessToast = (message, currentTheme) => {
  toast.success(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};

export const showWarningToast = (message, currentTheme) => {
  toast.warn(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};
