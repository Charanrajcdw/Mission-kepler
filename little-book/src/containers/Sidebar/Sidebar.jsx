import { useContext } from "react";
import { useSelector } from "react-redux";
import PropTypes from "prop-types";
import styles from "./Sidebar.module.scss";
import Checkbox from "../../components/Checkbox/Checkbox";
import { MODAL, SIDEBAR, THEME } from "../../constants";
import { ThemeContext } from "../../contexts/theme.context";

const Sidebar = ({ modalHandler }) => {
  const LOGO = SIDEBAR.logo.split(" ");
  const filterValues = useSelector((state) => state.blogs.blogTypes);
  const selectedBlogTypes = useSelector((state) => state.blogs.selectedBlogTypes);
  const { currentTheme, toggleTheme } = useContext(ThemeContext);

  const changeTheme = () => {
    toggleTheme();
  };

  const getFilterContent = () => {
    let filterContent = "";
    if (filterValues.length > 0) {
      filterContent = filterValues.map((filter) => (
        <Checkbox key={filter} label={filter} modalHandler={modalHandler} isSelected={selectedBlogTypes.includes(filter)} />
      ));
    } else {
      filterContent = <p>{SIDEBAR.noFilter}</p>;
    }
    return filterContent;
  };

  return (
    <aside className={styles.sidebarContainer}>
      <h1 className={styles.title}>
        {LOGO[0]} <span className={styles.titleLastWord}>{LOGO[1]}</span>
      </h1>
      <div className={styles.filtersContainer}>
        <p className={styles.filterTitle}>{SIDEBAR.filter}</p>
        <div>{getFilterContent()}</div>
      </div>
      <div className={styles.optionsContainer}>
        <p className={styles.members} onClick={() => modalHandler(MODAL.members, false)}>
          {SIDEBAR.viewMembers}
        </p>
        <p className={styles.theme} onClick={changeTheme}>
          {currentTheme === THEME.light ? SIDEBAR.switchToDark : SIDEBAR.switchToLight}
        </p>
      </div>
    </aside>
  );
};

Sidebar.propTypes = {
  modalHandler: PropTypes.func.isRequired,
};

export default Sidebar;
