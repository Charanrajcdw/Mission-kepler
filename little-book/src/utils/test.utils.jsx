import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import ThemeContextProvider from "../contexts/theme.context";
import blogSlice from "../store/slices/blogSlice";
import { THEME } from "../constants";

const customRender = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
    ...options
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return render(ui, { wrapper: ProviderWrapper, ...options });
};

const customCreateSnapshot = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return renderer.create(<ProviderWrapper>{ui}</ProviderWrapper>).toJSON();
};

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { customRender as render, customCreateSnapshot as createSnapshot };
