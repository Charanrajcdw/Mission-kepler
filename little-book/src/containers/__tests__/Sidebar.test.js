import { render, userEvent, screen, createSnapshot } from "../../utils/test.utils";
import Sidebar from "../Sidebar/Sidebar";
import { SIDEBAR, THEME } from "../../constants";

const blogTypes = ["type1", "type2", "type3"];
jest.mock("../../components/Checkbox/Checkbox", () => () => <div data-testid="checkbox" />);

describe("Sidebar UI", () => {
  it("render sidebar with filter values", () => {
    const initialBlogsData = { blogTypes: blogTypes };
    render(<Sidebar modalHandler={() => {}} />, { preloadedState: { blogs: initialBlogsData } });
    const logoElement = screen.getByRole("heading");
    const filterTitleElement = screen.getByText(SIDEBAR.filter);
    const checkboxElements = screen.getAllByTestId("checkbox");
    const membersElement = screen.getByText(SIDEBAR.viewMembers);
    const themeElement = screen.getByText(SIDEBAR.switchToDark);
    expect(logoElement).toHaveTextContent(SIDEBAR.logo);
    expect(checkboxElements).toHaveLength(3);
    expect(filterTitleElement).toBeInTheDocument();
    expect(membersElement).toBeInTheDocument();
    expect(themeElement).toBeInTheDocument();
  });

  it("render sidebar with no filter values", () => {
    render(<Sidebar modalHandler={() => {}} />);
    const logoElement = screen.getByRole("heading");
    const filterTitleElement = screen.getByText(SIDEBAR.filter);
    const membersElement = screen.getByText(SIDEBAR.viewMembers);
    const themeElement = screen.getByText(SIDEBAR.switchToDark);
    const noFilter = screen.getByText(SIDEBAR.noFilter);
    expect(logoElement).toHaveTextContent(SIDEBAR.logo);
    expect(noFilter).toBeInTheDocument();
    expect(filterTitleElement).toBeInTheDocument();
    expect(membersElement).toBeInTheDocument();
    expect(themeElement).toBeInTheDocument();
  });
});

describe("Sidebar functionalities", () => {
  it("should change theme on switch theme button click", async () => {
    const user = userEvent.setup();
    render(<Sidebar modalHandler={() => {}} />);
    const darkThemeElement = screen.getByText(SIDEBAR.switchToDark);
    expect(darkThemeElement).toBeInTheDocument();
    await user.click(darkThemeElement);
    const lightThemeElement = screen.getByText(SIDEBAR.switchToLight);
    expect(lightThemeElement).toBeInTheDocument();
  });

  it("should call modalHandler on view members button click", async () => {
    const user = userEvent.setup();
    const modalFn = jest.fn();
    render(<Sidebar modalHandler={modalFn} />);
    const darkThemeElement = screen.getByText(SIDEBAR.viewMembers);
    expect(darkThemeElement).toBeInTheDocument();
    await user.click(darkThemeElement);
    expect(modalFn).toHaveBeenCalled();
  });
});

describe("Sidebar snapshots", () => {
  it("should match sidebar snapshot with no filters", () => {
    const sidebar = createSnapshot(<Sidebar modalHandler={() => {}} />);
    expect(sidebar).toMatchSnapshot();
  });

  it("should match sidebar snapshot with filters", () => {
    const initialBlogsData = { blogTypes: blogTypes };
    const sidebar = createSnapshot(<Sidebar modalHandler={() => {}} />, { preloadedState: { blogs: initialBlogsData } });
    expect(sidebar).toMatchSnapshot();
  });
});
