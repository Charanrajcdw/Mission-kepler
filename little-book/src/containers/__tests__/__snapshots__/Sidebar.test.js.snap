// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar snapshots should match sidebar snapshot with filters 1`] = `
<aside
  className="sidebarContainer"
>
  <h1
    className="title"
  >
    Little
     
    <span
      className="titleLastWord"
    >
      Book
    </span>
  </h1>
  <div
    className="filtersContainer"
  >
    <p
      className="filterTitle"
    >
      FILTER
    </p>
    <div>
      <div
        data-testid="checkbox"
      />
      <div
        data-testid="checkbox"
      />
      <div
        data-testid="checkbox"
      />
    </div>
  </div>
  <div
    className="optionsContainer"
  >
    <p
      className="members"
      onClick={[Function]}
    >
      View Members
    </p>
    <p
      className="theme"
      onClick={[Function]}
    >
      Switch to Dark Mode
    </p>
  </div>
</aside>
`;

exports[`Sidebar snapshots should match sidebar snapshot with no filters 1`] = `
<aside
  className="sidebarContainer"
>
  <h1
    className="title"
  >
    Little
     
    <span
      className="titleLastWord"
    >
      Book
    </span>
  </h1>
  <div
    className="filtersContainer"
  >
    <p
      className="filterTitle"
    >
      FILTER
    </p>
    <div>
      <p>
        Add blogs to show filters
      </p>
    </div>
  </div>
  <div
    className="optionsContainer"
  >
    <p
      className="members"
      onClick={[Function]}
    >
      View Members
    </p>
    <p
      className="theme"
      onClick={[Function]}
    >
      Switch to Dark Mode
    </p>
  </div>
</aside>
`;
